Gems/Terrain/Code/Source/TerrainRaycast/TerrainRaycastContext.cpp

```diff
@@ -77,158 +77,167 @@ namespace
-    inline void FindNearestIntersectionIterative(const TerrainSystem& terrainSystem,
-                                                 const AZ::Vector2& terrainResolution,
-                                                 const AZ::Aabb& terrainWorldBounds,
-                                                 const AZ::Vector3& rayStart,
-                                                 const AZ::Vector3& rayEnd,
-                                                 AzFramework::RenderGeometry::RayResult& result)
+    /*
+       Iterative function that divides an AABB encompasing terrain points into grid squares based on
+       the given grid resolution and steps along the ray visiting each voxel it intersects in order
+       from nearest to farthest. In each square, it obtains the terrain height values at each corner and
+       triangulates them to find the nearest intersection (if any) between the triangles and the ray.
+
+       To step through the grid, we use an algorithm similar to Bresenham's line algorithm or a Digital
+       Differential Analyzer. We can't use Bresenham's line algorithm itself because it will sometimes skip
+       squares if the ray only passes through a tiny portion, and we need to use every square that it passes through.
+
+       We start by clipping the ray itself to the terrain AABB so that we don't walk through any grid squares
+       that cannot contain terrain. We then walk through the grid one square at a time, either moving horizontally
+       or vertically to the next square based on the ray's slope, until we reach the end of the ray or we've found a hit.
+    
+       Visualization:
+        - X: Grid square intersection but no triangle hit found
+        - T: Grid square intersection with a triangle hit found
+        ________________________________________
+        |    |    |    |    |    |    |    |    |
+        |____|____|____|____|____|____|____|____|  Ray
+        |    |    |    |    |    |    |    |    |  /
+        |____|____|____|____|____|____|____|____| /
+        |    |    |    |    |    |    |    | X  |/
+        |____|____|____|____|____|____|____|____/
+        |    |    |    |    |    |    |    | X /|
+        |____|____|____|____|____|____|____|__/_|
+        |    |    |    |    |    |    |    | /X |
+        |____|____|____|____|____|____|____|/___|
+        |    |    |    |    |    |    | X  / X  |
+        |____|____|____|____|____|____|___/|____|
+        |    |    |    |    |    |    | T/ |    |
+        |____|____|____|____|____|____|____|____|
+        |    |    |    |    |    |    |    |    |
+        |____|____|____|____|____|____|____|____|
+    */
+    inline void FindNearestIntersectionIterative(
+        const TerrainSystem& terrainSystem,
+        const AZ::Vector2& terrainResolution,
+        const AZ::Aabb& terrainWorldBounds,
+        const AZ::Vector3& rayStart,
+        const AZ::Vector3& rayEnd,
+        AzFramework::RenderGeometry::RayResult& result)
```

Comment:
suggestion: Maybe outside the scope of this PR but having this function actually return the result instead of using an out param might be a nice change 🙂

Reply:
or as this is in an anonymous namespace you can just omit entirely

Reply:
Good point, there's really no reason this even exists as a separate method, I'll merge the code into RayIntersect.